src/kernel/common.h

```diff
@@ -0,0 +1,108 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file kernel/common.h
+ * \brief Kernel common utilities
+ */
+#ifndef DGL_KERNEL_COMMON_H_
+#define DGL_KERNEL_COMMON_H_
+
+#include <dgl/runtime/ndarray.h>
+
+#include <cstdint>
+
+namespace dgl {
+namespace kernel {
+
+#ifdef __CUDACC__
+#define DGLDEVICE __device__
+#define DGLINLINE __forceinline__
+#else
+#define DGLDEVICE
+#define DGLINLINE inline
+#endif  // __CUDACC__
+
+#ifdef DGL_USE_CUDA
+#define DGL_XPU_SWITCH(val, Method, ...)  \
+  if (val == kDLCPU) {                    \
+    Method<kDLCPU>(__VA_ARGS__);          \
+  } else if (val == kDLGPU) {             \
+    Method<kDLGPU>(__VA_ARGS__);          \
+  } else {                                \
+    LOG(FATAL) << "Unsupported device type: " << val;  \
+  }
+#else  // DGL_USE_CUDA
+#define DGL_XPU_SWITCH(val, Method, ...)  \
+  if (val == kDLCPU) {                    \
+    Method<kDLCPU>(__VA_ARGS__);          \
+  } else {                                \
+    LOG(FATAL) << "Unsupported device type: " << val;  \
+  }
+#endif  // DGL_USE_CUDA
+
+#if 0
+#define GEN_DTYPE(GEN, ...)  \
+  GEN(__VA_ARGS__, float)    \
+  GEN(__VA_ARGS__, double)   \
+  GEN(__VA_ARGS__, int32_t)  \
+  GEN(__VA_ARGS__, int64_t)
+
+#define DGL_DTYPE_SWITCH(val, DType, ...)                   \
+  if (val.code == kDLInt && val.bits == 32) {               \
+    typedef int32_t DType;                                  \
+    {__VA_ARGS__}                                           \
+  } else if (val.code == kDLInt && val.bits == 64) {        \
+    typedef int64_t DType;                                  \
+    {__VA_ARGS__}                                           \
+  } else if (val.code == kDLFloat && val.bits == 32) {      \
+    typedef float DType;                                    \
+    {__VA_ARGS__}                                           \
+  } else if (val.code == kDLFloat && val.bits == 64) {      \
+    typedef double DType;                                   \
+    {__VA_ARGS__}                                           \
+  } else {                                                  \
+    LOG(FATAL) << "Unsupported dtype: " << val.code << "_"  \
+               << val.bits;                                 \
+  }
+#else
```

Comment:
So we only support float32?

Reply:
Yes. I plan to enable more data types later, but the compilation will become even slower ...

Reply:
Yes